### ISSUE TITLE
pnext: Support multiple IWPs connecting to data manager

### ISSUE THREAD

**Issue by eriohl** [Starcounter/Starcounter#2888]
Dynamic allocation of channels (and handle tables) needs to be supported.

**Comment by eriohl** [comment]
For this to work properly metadata table creation needs to be moved from IWP to the backend.

**Comment by alexanderstc** [comment]
> metadata table creation needs to be moved from IWP to the backend

Perhaps scdata can use existing scdbmetalayer.dll - just call `uint32_t star_create_metadata(uint64_t context_handle)`.

**Comment by eriohl** [comment]
It references sccoredb which is not available on scdata. So it is a bit more complicated then that. But possibly a specialised wp can be run doing this before other IWPs is allowed to run or something.

**Comment by eriohl** [comment]
@k-rus @alemoi @bigwad : I had an idea on this with metadata creation. How about we run this ```star_create_metadata``` as part of the build process to create a template for the log file. Then when creating a new database we can use this template to initialize the first log element file. Then metadata tables will already be there when initializing the database. What do you think?

**Comment by bigwad** [comment]
Why do we need such compilation? Is it because we need to run star_create_metadata in exclusive mode. like singleton initialization?
Then probably it's better to come up with some generic technic to handle such issues? Won't we come across the same problem on metadata update otherwise?

**Comment by eriohl** [comment]
> Why do we need such complication? Is it because we need to run star_create_metadata in exclusive mode, like singleton initialization? Then probably it's better to come up with some generic technic to handle such issues?

Yes. I'm trying to find a pragmatic way to make this work that does not involve rewriting a lot of code. I recognize it is not a beautiful solution. 

> Won't we come across the same problem on metadata update otherwise?

Well, the base metadata tables are kind of special. I think you have to basically recreate the database to change these anyway?

**Comment by k-rus** [comment]
I don't see any obstacles if metadata are created on build server. It also means that the creation of initial metadata cannot fail. This, of course, will complicate building process and also creation of databases. How would the meta-tables with meta-data appear in new databases? Will they be read from the log?



> Well, the base metadata tables are kind of special. I think you have to basically recreate the database to change these anyway?

Yes, currently it is required to start with an empty database and reapply DDLs to get correct state of metadata after meta-schema upgrade between Starcounter's versions.

**Comment by bigwad** [comment]
>Well, the base metadata tables are kind of special.

Am I right that the only special about base metatables is that kernel permits to create two layouts with the same token? Thus two metalayer initializations running in parallel may create two parallel meta-structures?
Then it's kind of kernel inconsistency: description of stari_context_create_layout says that layout token should be unique, but the kernel doesn't ensure this restriction in any way. Probably, it's enough for the kernel to reject creating layouts with the same token (probably with a special flag being passed to the stari_context_create_layout)? All other problems I can think of (such as retrying of creating base tables and filling them in exclusive mode) may be solved on the metalayer level.

And once proved, this technic may be extended to any client that requires exclusive database access.

**Comment by eriohl** [comment]
Actually: It is configurable weather the layout token (or index token) is enforced to be unique. This enforcement is currently turned of however. It was enforced before however, documentation needs to be updated.

The reason is is not enforced has to do with table upgrades in 2.3. All layouts for a specific table needs to have the same layout.

Regardless: Maybe you are on to something here. Maybe we just let the IWPs create the base metadata tables and use transaction conflict resolution to deal with multiple IWPs doing this simultaneously. Only solves one of the two problems related to this however (other problem being that related to asynchronous commits).

That is not what I meant with them being special however: What I meant is that there is no upgrade facility for them so you need to start from scratch when changing them.

**Comment by eriohl** [comment]
> How would the meta-tables with meta-data appear in new databases? Will they be read from the log?

Yes. The log is the database and the database is loaded into memory from the log always. It would be the same as loading the database for the second time is compared to now.

**Comment by bigwad** [comment]
>other problem being that related to asynchronous commits

I read something about it [here](https://github.com/Starcounter/Starcounter/issues/3582#issuecomment-220262364) but still don't get it (shame on me). Would you mind to go into some details on this?

**Comment by eriohl** [comment]
Tricky to explain better in this context but in essence:

1. ```star_context_commit_sync``` in synccommit.dll does won't work unless the host uses this library to do so.

2. ```star_create_metadata``` needs to commit data synchronously and currently it is using this library to do so (does not apply to any other metadata API however.)

3. We want use of synccommit.dll to be optional for the host.

4. And so: ```star_create_metadata``` can't use synccommit.dll.

5. Unless we don't call ```star_create_metadata``` from the host.

Not unsolvable of course but if we don't call ```star_create_metadata``` from the IWP then is kind of solves it self without any extra effort on our part.

**Comment by bigwad** [comment]
Thanks, I see now.
And the reason for №2 is that star_create_metadata split it work into several transactions and the reason for that is that kernel doesn't allow creating and working with a layout in the single transaction. Right?
So the answer could be splitting star_create_metadata into several methods (probably, using continuation-passing?).

More generally, the problem is that async commits require all work to be done on host transaction as only the host knows how to handle notifications.

Two question here:
1. Isn't it going to be a problem with the upcoming mapping anyway? As I heard mapping may restructure underlying database a lot. Is it possible that this restriction hits it?
2. This applies to other metalayer functions as well. So metalayer should stop creating its own transactions and start requiring host transactions. Right?

**Comment by k-rus** [comment]
> More generally, the problem is that async commits require all work to be done on host transaction as only the host knows how to handle notifications.



> 2. This applies to other metalayer functions as well. So metalayer should stop creating its own transactions and start requiring host transactions. Right?

The host must not expose ability to create transactions to an application code in the case of DDL calls, because of layout creation cannot be combined with layout accesses as you mentioned above. To guarantee this the metalayer does not allow host transactions. Furthermore, the metalayer controls if a transaction succeeds or fails for query cache invalidation. However, if the kernel limitation will be removed in future, we should reconsider the current design.



> 1. Isn't it going to be a problem with the upcoming mapping anyway? As I heard mapping may restructure underlying database a lot. Is it possible that this restriction hits it?

No, the mapper will not affect the metalayer schema, which is internal and cannot be modified outside the metalayer. The mapper works with user types.

**Comment by bigwad** [comment]
>To guarantee this the metalayer does not allow host transactions. Furthermore, the metalayer controls if a transaction succeeds or fails for query cache invalidation. 

As I guessed, metalayer (and any other entity except the host itself) can't commit transactions. If it's true, then metalayer-host interaction should be redesigned.

>No, the mapper will not affect the metalayer schema, which is internal and cannot be modified outside the metalayer. The mapper works with user types.

I don't consider solely metalayer data here, but any changes that are required to be applied to the database to support mapping. If mapping doesn't require any - that's fine.

**Comment by eriohl** [comment]
The idea discussed is basically to treat the DDL operations them selves as an asynchronous operation. This then allows them to create the, and initialize commit of, the transaction. Although perhaps not be able to determine the result of the operation. It was then assumed that this worked fine for every operation except for ```star_create_metadata```. As discussed with @alexanderstc previously.

If that is not true as implied (that is, if the commit needs to be synchronous for all DDL operations because the DDL operation needs to know the result of the commit) then this doesn't work as discussed and we need to rethink this. Metalayer can not be dependent on synccommit as is it now in future.

Anyway: We don't support asynchronous commits anyway now and we will not for a while since it seems not to be much of a priority anyway since it has been determined that 2.3 will continue using the legacy scheduler and so we will be unable to implement this anyway. So I propose resolving this is not to acute.

**Comment by alexanderstc** [comment]
> Metalayer can not be dependent on synccommit 

I'll try to program at least the sql cache invalidation without relying on knowledge of completion of a commit.

@k-rus should evaluate if we can program the rest of DDL with async commits.

**Comment by bigwad** [comment]
> If that is not true as implied (that is, if the commit needs to be synchronous for all DDL operations because the DDL operation needs to know the result of the commit) then this doesn't work as discussed and we need to rethink this. Metalayer can not be dependent on synccommit as is it now in future.

I suppose that's twofold:
1. User code is likely to be interested in DDL result (at least to get an exception on the conflict).
2. A Subsequent transaction is probably dependent on DDL, implying that DDL should be synchronous.

**Comment by bigwad** [comment]
>Anyway: We don't support asynchronous commits anyway now and we will not for a while since it seems not to be much of a priority anyway since it has been determined that 2.3 will continue using the legacy scheduler and so we will be unable to implement this anyway. So I propose resolving this is not to acute.

My point here is that we have more problems with sync commits, than just metalyer schema initialization.

**Comment by eriohl** [comment]
> I suppose that's twofold:
> 1. User code is likely to be interested in DDL result (at least to get an exception on the conflict).
> 2. A subsequent transaction is probably dependent on DDL, implying that DDL should be synchronous.

We'll obviously a transaction, or any thing else (like returning the result to the client), relying on the result of a previous transaction must execute after the previous transaction is completed. This is true for any transaction.

But: How this is handled must be up to the host. In Node.js for example is would have to be a callback when transaction is completed. Same goes for .NET with async/await (although from the programmers point of view it won't look like a callback). In a more traditional approach the thread will block waiting for the operation to complete. How this is implemented must however be up the host. Which is why lower level operations will have a hard time relying on synchronous commits.

But anyway: Maybe we are talking around each other on this one. Let's table this discussion until we can have it work it out in a better setting.

**Comment by k-rus** [comment]
> As I guessed, metalayer (and any other entity except the host itself) can't commit transactions.

I don't really understand why you guessed so. Metalayer creates and commits transactions, since DDLs do not involve any application or host transactional code.

> Metalayer can not be dependent on synccommit
> I'll try to program at least the sql cache invalidation without relying on knowledge of completion of a commit.
> 
> @k-rus should evaluate if we can program the rest of DDL with async commits.

DDLs implementation uses single transaction, thus it does not need to do synchronous call, especially if the query cache is designed not to control the commits.

I agree it will be correct to move transaction outside metalayer into a host. But then host needs to create internal transaction on DDL calls, which includes DDLs related to class definitions and DDLs through SQL (e.g. insert statements), and fail if DDL calls (including DDL SQL) comes in a user transaction. So it becomes not so simple and host needs knowledge about different SQL types.

**Comment by bigwad** [comment]
>>As I guessed, metalayer (and any other entity except the host itself) can't commit transactions.

> I don't really understand why you guessed so. Metalayer creates and commits transactions, since DDLs do not involve any application or host transactional code.

First, I must reiterate that here we're talking about hosts with async commit model.

As I mentioned before, I see two points that make it impossible to make async commits from DDL:
1. User code is likely to be interested in DDL result (at least to get an exception on the conflict).
2. A subsequent transaction is probably dependent on DDL, implying that DDL should be synchronous.

Actually, the second item is too strict and don't say what I actually was trying to say. I meant that host should have a means to ensure DDL is finished until it may continue executing code that depends on that DDL. Thus DDL should be either synchronous or adapt host's async model.

**Comment by k-rus** [comment]
> 1. User code is likely to be interested in DDL result (at least to get an exception on the conflict).

This is not fully clear for me if it will be so. The only DDL, which user code can contain nowadays, is create index statement. The rest is done internally in a host.



> 2. A subsequent transaction is probably dependent on DDL, implying that DDL should be synchronous.
> 
> Actually, the second item is too strict and doesn't say what I actually was trying to say. I meant that host should have a means to ensure DDL is finished until it may continue executing code that depends on that DDL. Thus DDL should be either synchronous or adapt host's async model.

Thus I agreed with the correctness of moving transaction handling to hosts, but the described complexity should be taken in account. So none of the solutions are clean.

**Comment by bigwad** [comment]
The simplest solution as I see is to provide "native" blocking synchronous commit, that doesn't interact with host's transaction model. It will reduce throughput if host supports async commits, but I think it's acceptable for DDL and metalayer initialization.

**Comment by alexanderstc** [comment]
@bigwad sounds nice and simple.
@eriohl would you approve that?

**Comment by eriohl** [comment]
> The simplest solution as I see is to provide "native" blocking synchronous commit, that doesn't interact with host's transaction model. It will reduce throughput if host supports async commits, but I think it's acceptable for DDL and metalayer initialization.

I don't follow what you mean by "native" blocking synchronous commit?

But I think that in general there should be no kernel operations blocking the thread for the simple reason that the threading model of the host may not allow this (like Node.js for example). So it doesn't matter if you ask the host to block the thread or support kernel blocking the thread because if might simply not be possible to block the thread.

**Comment by bigwad** [comment]
By "native" I mean that kernel provides sync commit that doesn't interact with host's commit model (practically: use another pipe for sync notification or simply blocks inside kernel channel). 
Don't see how it may clash with the host's threading model. It's not perfectly aligned with async hosts design (like node.js), but still legitimate. Even official node.js API has several sync calls (for file IO for example).

**Comment by eriohl** [comment]
Well, regardless of that: Surely we can agree that it would be better if there was no synchronous commits (or other blocking operations) in the kernel of metadata and this is what we should strive for.

**Comment by bigwad** [comment]
Recently I've made some work regarding multiple commits in the metalayer. Main idea is to rewrite functions that perform multiple commits using continuation passing style, where each continuation is supposed to commit part of the job. Part of this work is that metalayer now operates on external transactions exclusively - that means no transactions in managed code. Host code is updated to create transaction when needed. Though no changes to user-faced behaviour: it's still illegal to create an index inside transaction scope.

@alexanderstc, @k-rus, It's kind of major refactoring to the metalayer so I urge you guys to review and (dis)approve the changes so I can move on. Code is in the **issue/3582** branches for both level0 and level1.

**Comment by alexanderstc** [comment]
@bigwad could you please merge pnext into both your branches.
After conflict resolution it will be easier for me to see how your l0 branch differs from the current pnext.

I also assume your branches will pass buildserver test.

**Comment by bigwad** [comment]
I merged it yesterday evening. I'll merge it again shortly (seems a little conflict was introduced).
And yes, it passes tests.

**Comment by bigwad** [comment]
Merged. There is only one minor change. Passes my local tests but can't push as gitlab is down.

**Comment by alexanderstc** [comment]
@k-rus @bigwad 

I'd like you to consider also my idea about imitating transaction-scoped DDL operations in our database by pooling kernel objects - tables and indices - beforehand, and, when a user require a table and/or index to be created, take required object(s) from the pool and update metadata accordingly.

That approach may produce a more ACID-looking DDL with not so much of programming effort.

**Comment by bigwad** [comment]
Pushed

**Comment by k-rus** [comment]
> I'd like you to consider also my idea about imitating transaction-scoped DDL operations in our database by pooling kernel objects - tables and indices - beforehand, and, when a user require a table and/or index to be created, taking required object(s) from the pool and updating metadata accordingly.

I think it's interesting idea. @bigwad @eriohl WDYT

**Comment by alexanderstc** [comment]
> I'd like you to consider also my idea about imitating transaction-scoped DDL operations in our database by pooling kernel objects - tables and indices - beforehand, and, when a user require a table and/or index to be created, taking required object(s) from the pool and updating metadata accordingly.

I believe I could implement the whole thing inside the metalayer, level0, entirely.

It would lack CPS code though, which would make me, a scheme hobbyist, kind of sad.

**Comment by bigwad** [comment]
>I'd like you to consider also my idea about imitating transaction-scoped DDL operations

It's interesting and I'd like to get more details, but is it in any way crossing with metalayer's asynchronous commits?

**Comment by alexanderstc** [comment]
The very first comment in this thread:

> For this to work properly metadata table creation needs to be moved from IWP to the backend.

Well, _metadata, table creation_, can it be relevant somehow?
Or maybe I'm just reaching and trying to push my pet idea.

**Comment by alexanderstc** [comment]
Object pooling may enable user-level DDL transaction control, no?

**Comment by bigwad** [comment]
>Object pooling may enable user-level DDL transaction control, no?

It could. Depends on implementation.


My point is following - I have to move on with asynchronous transactions. We had a problem with that inside the metalayer, that could have been fixed in several ways. I've implemented solution that:
 1. doesn't require to move metalayer initialization elsewhere.
 2. establishes CPS helpers should need in waiting for commit arises in some other metalayer functions.
 3. puts DDL transactions under host (not user) control.

As I said, I need kind of such solution to move on. If you propose to move DDL transaction control further to the user, it's fine. What I want to know now - does my solution make any problem for your future work or I can merge it?

**Comment by alexanderstc** [comment]
> does my solution make any problem

No problem for me.

**Comment by bigwad** [comment]
By the way, I hope, finally we get [this](https://git.starcounter.org/BlueStar/level0/issues/85) fixed. Marrying TLS and CPS was literally a nightmare. And I didn't even try to run continuation on a different thread.

**Comment by alexanderstc** [comment]
@bigwad 
do I understand correctly that the metalayer doesn't create/commit/free transactions anymore?

**Comment by bigwad** [comment]
Right. 
I wanted to move Transaction class from the metalayer to make that clear, but there are some interdependencies between tests, Transaction class, KernelContext class, so I left it in metalayer.

**Comment by alexanderstc** [comment]
@bigwad 

> Marrying TLS and CPS was literally a nightmare

You should've let me help.
Together we'd wasted less time, effort, and quite possibly, made a bit cleaner code.

**Comment by bigwad** [comment]
>You should've let me help.
Together we'd wasted less time, effort, and quite possibly, made a bit cleaner code.

Thanks, the marrying process was primarily in my mind :smile:. Surprisingly little amount of code has been actually written about that.
But probably, you're right, it could be cleaner. But I didn't want to put too much effort into it if we have a plan to get rid of TLS eventually.

**Comment by alexanderstc** [comment]
@bigwad please don't commit your change to pnext for the moment.

**Comment by alexanderstc** [comment]
@bigwad you are again welcome to push to pnext.

**Comment by alexanderstc** [comment]
See also https://git.starcounter.org/BlueStar/level0/issues/153
